industrial_taxonomy/pipeline/NSPL/utils.py

```diff
@@ -0,0 +1,141 @@
+"""Helper functions for NSPL flow."""
+import re
+from io import BytesIO
+from typing import Optional
+from zipfile import ZipFile
+
+import pandas as pd
+import requests
+from selenium import webdriver
+from selenium.webdriver.chrome.options import Options
+from selenium.webdriver.chrome.webdriver import WebDriver
+import industrial_taxonomy
+
+LOOSE_UK_BOUNDS = {
+    "long": (-7.6, 1.7),
+    "lat": (50.0, 58.7),
+}  # Rough bounds of UK, ok for DQ check
+#
+
+
+def find_download_url(driver: WebDriver, geo_portal_url: str) -> str:
+    """Find download button and extract download URL."""
+    driver.implicitly_wait(10)  # Poll DOM for up to 10 seconds when finding elements
+    driver.get(geo_portal_url)
+    return driver.find_element_by_link_text("Download").get_attribute("href")
+
+
+#
+def chrome_driver() -> WebDriver:
+    """Headless Selenium Chrome Driver."""
+    chrome_options = Options()
+    chrome_options.add_argument("--headless")
+    chrome_options.add_argument("--no-sandbox")
+    chrome_options.add_argument("--disable-gpu")
+    driver = webdriver.Chrome(
+        executable_path="/Users/imanmuse/miniconda3/chromedriver",
+        options=chrome_options,
+    )
+    driver.set_page_load_timeout(10)
+    driver.set_script_timeout(10)
+    return driver
+
+
+#
+
+
+def filter_nspl_data(data: pd.DataFrame) -> pd.DataFrame:
+    """Filter out NSPL rows without location data."""
+    keep_rows = (
+        data.laua_code.notna()
```

Comment:
Reading through the error trace I get when running, it's clear that this is where the error arises.

> 2022-01-14 09:03:26.042 [1642150885703686/get_data/2 (pid 335866)] File "/home/alex/GIT/industrial_taxonomy/industrial_taxonomy/pipeline/NSPL/flow.py", line 104, in get_data
2022-01-14 09:03:26.042 [1642150885703686/get_data/2 (pid 335866)] self.nspl_data = read_nspl_data(zipfile, nrows).pipe(filter_nspl_data)
2022-01-14 09:03:26.042 [1642150885703686/get_data/2 (pid 335866)] File "/home/alex/anaconda3/envs/industrial_taxonomy/lib/python3.8/site-packages/pandas/core/generic.py", line 5430, in pipe
2022-01-14 09:03:26.042 [1642150885703686/get_data/2 (pid 335866)] return com.pipe(self, func, args, kwargs)
2022-01-14 09:03:26.042 [1642150885703686/get_data/2 (pid 335866)] File "/home/alex/anaconda3/envs/industrial_taxonomy/lib/python3.8/site-packages/pandas/core/common.py", line 471, in pipe
2022-01-14 09:03:26.042 [1642150885703686/get_data/2 (pid 335866)] return func(obj, args, kwargs)
2022-01-14 09:03:26.043 [1642150885703686/get_data/2 (pid 335866)] File "/home/alex/GIT/industrial_taxonomy/industrial_taxonomy/pipeline/NSPL/utils.py", line 50, in filter_nspl_data
2022-01-14 09:03:26.043 [1642150885703686/get_data/2 (pid 335866)] data.laua_code.notna()
2022-01-14 09:03:26.043 [1642150885703686/get_data/2 (pid 335866)] File "/home/alex/anaconda3/envs/industrial_taxonomy/lib/python3.8/site-packages/pandas/core/generic.py", line 5487, in __getattr__
2022-01-14 09:03:26.043 [1642150885703686/get_data/2 (pid 335866)] return object.__getattribute__(self, name)
2022-01-14 09:03:26.043 [1642150885703686/get_data/2 (pid 335866)] AttributeError: 'DataFrame' object has no attribute 'laua_code'


This is happening because you modified the `read_nspl_data` function to set `laua_code` as the index which means it is not accessible as a column on the `DataFrame` any longer.

Reply:
Thank you for that!! It really helped! I now have the metaflow running successfully and commited my most recent draft.   https://github.com/nestauk/industrial_taxonomy/pull/28/commits/959f7e6d811e217776bafa8998a08e55833aba8e . I did have issues with the data_quaility step, but I have commented it out for now.